docs/developer_overview.md

```diff
@@ -0,0 +1,511 @@
+# CUB Developer Overview
+
+
+This living document serves as a guide to the design of the internal structure of CUB. 
+
+The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 
+
+These layers are:
+- Thread 
+- Warp
+- Block
+- Device
+
+For example, consider `reduce`. CUB has four flavors of `reduce` that correspond to these different layers:
+
+
+| layer                | coop invocation | coop execution | max threads | valid result in |
+| -------------------- | --------------- | -------------- | ----------- | --------------- |
+| `cub::ThreadReduce`  | -               | -              | 1           | invoking thread |
+| `cub::WarpReduce`    | +               | +              | 32          | main thread     |
+| `cub::BlockReduce`   | +               | +              | 1024        | main thread     |
+| `cub::DeviceReduce ` | -               | +              | ∞           | global memory   |
+
+## Thread-scope
+
+Thread-scope functionality is usually presented with functions and rarely exposed to the user:
+
```

Comment:
Let's avoid any ambiguity with `cuda::thread_scope`s and just avoid the term "scope" all together. 

```suggestion
## Single Thread

Single threaded functionality like `cub::ThreadReduce` is typically implemented as a sequential function and rarely exposed to the user. 

```